src/bm_sim/match_units.cpp

```diff
@@ -584,6 +590,7 @@ MatchKeyBuilder::match_params_sanity_check(
           return false;
         break;
       case MatchKeyParam::Type::TERNARY:
+      case MatchKeyParam::Type::OPTIONAL:
```

Comment:
let's treat them differently here. This function (sanity check) is probably a great place to validate the mask for the optional case.

Reply:
one more thing that comes to mind:
the check `if (param.type != f_info.mtype) return false` above will fail at the moment for optional match fields, that will require changes in src/bm_runtime